contracts/Utils.sol

```diff
@@ -9,9 +9,27 @@ contract Utils {
 
     ERC20 constant internal ETH_TOKEN_ADDRESS = ERC20(0x00eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee);
     uint  constant internal PRECISION = (10**18);
-    uint  constant internal MAX_QTY   = (10**28); // 1B tokens
+    uint  constant internal MAX_QTY   = (10**28); // 10B tokens
     uint  constant internal MAX_RATE  = (PRECISION * 10**6); // up to 1M tokens per ETH
     uint  constant internal MAX_DECIMALS = 18;
+    uint  constant internal ETH_DECIMALS = 18;
+    mapping(address=>uint) internal decimals;
+
+    function setDecimals(ERC20 token) internal {
+        if (token == ETH_TOKEN_ADDRESS) decimals[token] = ETH_DECIMALS;
+        else decimals[token] = token.decimals();
+    }
+
+    function getDecimals(ERC20 token) internal view returns(uint) {
+        if (token == ETH_TOKEN_ADDRESS) return ETH_DECIMALS; // save storage access
+        uint tokenDecimals = decimals[token];
+        // technically, there might be token with decimals 0
+        // moreover, very possible that old tokens have decimals 0
+        // these tokens will just have higher gas fees.
+        if(tokenDecimals == 0) return token.decimals();
+
```

Comment:
I suggest if (tokenDecimals == 0) 
set decimals[token]
this way can avoid setDecimals all together. don't have to define it or call it.
a bit cleaner code imo.

Reply:
Thought of that, but then certain functions are no longer constant.

Reply:
I see your point.
when ever calling a non constant function it means it can't be a local read?

Reply:
well, the sad thing is that technically it can, because solidity compiler does not verify it. but you get compilation warnings.
So it is not a good practice.